hollow/src/main/java/com/netflix/hollow/tools/history/keyindex/HollowHistoryTypeKeyIndex.java

```diff
@@ -18,485 +18,243 @@
-                return false;
-        }
-        return true;
-    }
-
-    private boolean recordFieldMatchesKey(HollowObjectTypeReadState typeState, int ordinal, HollowObjectTypeReadState keyTypeState, int keyOrdinal, int keyFieldPosition, String[] keyFieldParts, int keyFieldPartPosition) {
-        int schemaPosition = typeState.getSchema().getPosition(keyFieldParts[keyFieldPartPosition]);
-        if(keyFieldPartPosition < keyFieldParts.length - 1) {
-            HollowObjectTypeReadState nextPartTypeState = (HollowObjectTypeReadState) typeState.getSchema().getReferencedTypeState(schemaPosition);
-            int nextOrdinal = typeState.readOrdinal(ordinal, schemaPosition);
-            return recordFieldMatchesKey(nextPartTypeState, nextOrdinal, keyTypeState, keyOrdinal, keyFieldPosition, keyFieldParts, keyFieldPartPosition + 1);
-        } else {
-            return HollowReadFieldUtils.fieldsAreEqual(typeState, ordinal, schemaPosition, keyTypeState, keyOrdinal, keyFieldPosition);
-        }
-    }
-
-    private void copyExistingKeys() {
-        HollowTypeWriteState typeState = writeStateEngine.getTypeState(primaryKey.getType());
-        if (typeState == null) return;
-
-        typeState.addAllObjectsFromPreviousCycle();
-    }
-
-    private void populateAllCurrentRecordKeysIntoIndex(HollowObjectTypeReadState typeState) {
-        HollowObjectWriteRecord rec = new HollowObjectWriteRecord(keySchema);
-        PopulatedOrdinalListener listener = typeState.getListener(PopulatedOrdinalListener.class);
-        BitSet previousOrdinals = listener.getPreviousOrdinals();
-        BitSet populatedOrdinals = listener.getPopulatedOrdinals();
-
-        int maxLength = Math.max(previousOrdinals.length(), populatedOrdinals.length());
-
-        for(int i=0;i<maxLength;i++) {
-            if(populatedOrdinals.get(i) || previousOrdinals.get(i))
-                writeKeyObject(typeState, i, rec);
-        }
-    }
-
-    private void populateNewCurrentRecordKeysIntoIndex(HollowObjectTypeReadState typeState) {
-        HollowObjectWriteRecord rec = new HollowObjectWriteRecord(keySchema);
-        PopulatedOrdinalListener listener = typeState.getListener(PopulatedOrdinalListener.class);
-        BitSet populatedOrdinals = listener.getPopulatedOrdinals();
-        BitSet previousOrdinals = listener.getPreviousOrdinals();
-
-        RemovedOrdinalIterator iter = new RemovedOrdinalIterator(populatedOrdinals, previousOrdinals);
-        int ordinal = iter.next();
-        while(ordinal != ORDINAL_NONE) {
-            writeKeyObject(typeState, ordinal, rec);
-            ordinal = iter.next();
-        }
-    }
-
-    public String[] getKeyFields() {
-        return primaryKey.getFieldPaths();
+    public void addMatches(int hashCode, IntList results) {
+        if (!ordinalFieldHashMapping.containsKey(hashCode))
+            return;
+        IntList res2 = ordinalFieldHashMapping.get(hashCode);
```

Comment:
A little bug here in that the query returns results based on hash match rather than value match. This causes the search behavior to sometimes return garbage results.

Reply:
Fixed in faad5249